torchqdynamics/ode/adjoint.py

```diff
@@ -0,0 +1,110 @@
+import torch
+import torch.nn as nn
+import torchqdynamics.ode as ode
+
+# -------------------------------------------------------------------------------------------------
+#     Adjoint integrator base class
+# -------------------------------------------------------------------------------------------------
+
+class ODEIntAdjoint(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, f, f_adj, y0, tspan, params, solver='dopri5', save_at=(), atol=1e-8, rtol=1e-6):
+        # Save into context for backward pass
+        ctx.f = f
+        ctx.f_adj = f_adj
+        ctx.tspan = tspan
+        ctx.solver = solver
+        ctx.atol = atol
+        ctx.rtol = rtol
+
+        # Solve the ODE forward and save results
+        with torch.no_grad():
+            t, y = ode.odeint(f, y0, tspan, solver, save_at, atol=atol, rtol=rtol)
+            ctx.save_for_backward(t, y)
+        return t, y
+
+    @staticmethod
+    def backward(ctx, *grad_y):
+        # Disable gradient computation
+        torch.set_grad_enabled(False)
+        
+        # Unpack context
+        f = ctx.f
+        f_adj = ctx.f_adj
+        tspan = ctx.tspan
+        solver = ctx.solver
+        atol = ctx.atol
+        rtol = ctx.rtol
+        t_saved, y_saved = ctx.saved_tensors
+
+        # Get model parameters
+        params = torch.cat([p.view(-1) for p in f.parameters() if p.requires_grad])
+        params_ = tuple(f.parameters())
+        
+        # Initialize the augmented state as a flat tensor
+        # y is the state, a is the adjoint, g is the gradient w.r.t. parameters
+        yT, aT, gT = y_saved[-1], grad_y[-1][-1], torch.zeros_like(params)
+        y_len, a_len, g_len = yT.numel(), aT.numel(), gT.numel()
+        y_shape, a_shape, g_shape = yT.shape, aT.shape, gT.shape
+        aug_y = torch.cat([yT.view(-1), aT.view(-1), gT.view(-1)])
+
+        solver_type = ode.solver_to_solvertype(solver)
+        if solver_type == "rk":
```

Comment:
Better declare an enumeration here 
```
class Solver:
    RK = "rk"
    ROUCHON = "rouchon"
    [etc...]
```
and then later we can refactor easily and not be stuck with strings

Reply:
I'm not familiar with that. How could I implement that in this particular exemple? Like, where are the two functions actually defined?

Reply:
See abocquet/odeint-enums